### PR TITLE
Hooks separate actions from entry catalog

### PR DESCRIPTION
**Pull request by nopcoder**

### CODE REVIEW

File: actions/webhook.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	"time"
+)
+
+type Webhook struct {
+	ID         string
+	ActionName string
+	URL        string
+}
+
+type WebhookEventInfo struct {
```
Comment (itaiad200) [inline actions/webhook.go:19]: I don't like that our API with customers is being handled this way. It's so easy to break it without ever knowing.

File: actions/action.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	default:
+		return false, nil
+	}
+	// action without branches spec is matched
+	if len(actionOn.Branches) == 0 {
+		return true, nil
+	}
+	// find at least one match
+	for _, b := range actionOn.Branches {
+		matched, err := path.Match(b, spec.Branch)
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply actions/action.go:90]: wanted something that I can reference and reuse - path.Match is like fnmatch which is very close to what github actions supports

File: actions/webhook.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	// log response body if needed
+	if resp.Body != nil && resp.ContentLength > 0 {
+		if err := writer.OutputWrite(ctx, w.ID, resp.Body); err != nil {
+			return err
+		}
+	}
+
+	// check status code
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		return fmt.Errorf("%w (%d)", ErrWebhookResponseStatusCode, resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply actions/webhook.go:73]: I'll remove the ErrXXX but I still need to specify which hook failed - something in the lines of:
webhook request fail with status code: %d

File: catalog/entry_catalog.go [hunk 4]
```diff
@@ -505,9 +513,27 @@ func (e *EntryCatalog) DumpTags(ctx context.Context, repositoryID graveler.Repos
 }
 
 func (e *EntryCatalog) preCommitHook(ctx context.Context, repositoryID graveler.RepositoryID, branchID graveler.BranchID, commit graveler.Commit) error {
+	_ = actions.Event{
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply catalog/entry_catalog.go:516]: currently, this code does nothing - just here to capture what data we need to pass to 'actions' on each event.

File: actions/webhook.go (hunk 1 above)
Comment (nopcoder) [reply actions/webhook.go:19]: For now, I'll keep it as is, until we will have to address compatibility changes and need to address versioning.

File: actions/webhook.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nopcoder) [reply actions/webhook.go:73]: Can't format dynamic error - not recommended. I'll produce the above using ErrXXX

File: actions/webhook.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	eventData, err := w.marshalEvent(runID, ed)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest(http.MethodPost, w.URL, bytes.NewReader(eventData))
+	if err != nil {
+		return err
+	}
+	client := &http.Client{
+		Timeout: time.Minute,
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply actions/webhook.go:54]: const default and optional override by hook's properties

File: catalog/entry_catalog.go [hunk 6]
```diff
@@ -5,8 +5,12 @@ import (
 	"crypto"
 	_ "crypto/sha256"
 	"fmt"
+	"time"
+
+	"github.com/treeverse/lakefs/actions"
```
Comment (itaiad200) [inline catalog/entry_catalog.go:11]: dd

Comment (nopcoder) [reply catalog/entry_catalog.go:11]: but I'm using the constants for the event type from actions

File: catalog/entry_catalog.go [hunk 7]
```diff
@@ -129,7 +134,10 @@ func NewEntryCatalog(cfg *config.Config, db db.Database) (*EntryCatalog, error)
 	refManager := ref.NewPGRefManager(db, ident.NewHexAddressProvider())
 	branchLocker := ref.NewBranchLocker(db)
 	store := graveler.NewGraveler(branchLocker, committedManager, stagingManager, refManager)
-	entryCatalog := &EntryCatalog{Store: store}
+	entryCatalog := &EntryCatalog{
+		BlockAdapter: tierFSParams.Adapter,
```
Comment (itaiad200) [inline catalog/entry_catalog.go:138]: TierFSParams??

File: actions/hook.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+	OutputWrite(ctx context.Context, name string, reader io.Reader) error
+}
+
+type Hook interface {
+	Run(ctx context.Context, runID string, event Event, writer OutputWriter) error
+}
+
+type NewHookFunc func(*Action, ActionHook) Hook
+
+var hooks = map[HookType]NewHookFunc{
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply actions/hook.go:26]: not committed?

**Review by itaiad200 — APPROVED** [review]
Nice! just a couple of small comments remain

File: catalog/entry_catalog.go (hunk 7 above)
Comment (nopcoder) [reply catalog/entry_catalog.go:138]: Result of GetCommittedTierFSParams includes the configured block adapter used by the storage.

**Pull request merged**